File: 4. Estructuras de datos lineales/Python/pilas.py
# 1. Implementa una pila con funciones para apilar, desapilar y mostrar el elemento superior.


class Pila:
    def __init__(self):
        self.pila = []

    def apilar(self, elemento):
        self.pila.append(elemento)

    def desapilar(self):
        if not self.esta_vacia():
            return self.pila.pop()
        return None

    def mostrar_superior(self):
        if not self.esta_vacia():
            return self.pila[-1]
        return None

    def esta_vacia(self):
        return len(self.pila) == 0

    def __str__(self):
        return str(self.pila)


pila = Pila()
print(pila)
pila.apilar(1)
pila.apilar(2)
pila.apilar(3)
print(pila)


# 2. Escribe un programa que invierta el contenido de una cadena usando una pila.


def invertir_cadena(cadena):
    pila = Pila()
    for caracter in cadena:
        pila.apilar(caracter)
    cadena_invertida = ""
    while not pila.esta_vacia():
        cadena_invertida += pila.desapilar()
    return cadena_invertida


cadena = "Hola mundo"
print(cadena)
print(invertir_cadena(cadena))

File: 4. Estructuras de datos lineales/JavaScript/pilas.js
// clase Pila
class Pila {
  constructor() {
    this.pila = [];
  }
  push(elemento) {
    this.pila.push(elemento);
  }
  pop() {
    return this.pila.pop();
  }
  peek() {
    return this.pila[this.pila.length - 1];
  }
  isEmpty() {
    return this.pila.length === 0;
  }
  size() {
    return this.pila.length;
  }
  print() {
    console.log(this.pila);
  }
}

// 1. Implementa una pila que permita verificar si una cadena tiene paréntesis balanceados.
function parentesisBalanceados(cadena) {
  const pila = new Pila();
  for (let i = 0; i < cadena.length; i++) {
    if (cadena[i] === "(") {
      pila.push(cadena[i]);
    } else if (cadena[i] === ")") {
      if (pila.isEmpty()) {
        return false;
      }
      pila.pop();
    }
  }
  return pila.isEmpty();
}

console.log(
  "La cadena '(())' tiene paréntesis balanceados: ",
  parentesisBalanceados("(())")
); // true
console.log(
  "La cadena '(()' tiene paréntesis balanceados: ",
  parentesisBalanceados("(()")
); // false

// 2. Escribe una función que convierta un número decimal a binario usando una pila.
function decimalToBinary(decimal) {
  const pila = new Pila();
  let binario = "";
  while (decimal > 0) {
    pila.push(decimal % 2);
    decimal = Math.floor(decimal / 2);
  }
  while (!pila.isEmpty()) {
    binario += pila.pop();
  }
  return binario;
}

console.log("10 en binario es: ", decimalToBinary(10)); // 1010
console.log("15 en binario es: ", decimalToBinary(15)); // 1111
